File: src/main/java/at/jku/ssw/coco/intellij/structure/CocoStructureViewFactory.java
package at.jku.ssw.coco.intellij.structure;

import at.jku.ssw.coco.intellij.psi.CocoFile;
import com.intellij.ide.structureView.StructureViewBuilder;
import com.intellij.ide.structureView.StructureViewModel;
import com.intellij.ide.structureView.TreeBasedStructureViewBuilder;
import com.intellij.lang.PsiStructureViewFactory;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by Thomas on 29/03/2015.
 */
public class CocoStructureViewFactory implements PsiStructureViewFactory {
    @Nullable
    @Override
    public StructureViewBuilder getStructureViewBuilder(PsiFile psiFile) {
        if (!(psiFile instanceof CocoFile)) {
            return null;
        }

        return new TreeBasedStructureViewBuilder() {
            @NotNull
            @Override
            public StructureViewModel createStructureViewModel(@Nullable Editor editor) {
                return new CocoStructureViewModel((CocoFile) psiFile, editor);
            }

            @Override
            public boolean isRootNodeShown() {
                return false;
            }
        };
    }
}

File: src/main/kotlin/me/salzinger/intellij/coco/structure/CocoStructureViewFactory.kt
package me.salzinger.intellij.coco.structure

import com.intellij.ide.structureView.StructureViewBuilder
import com.intellij.ide.structureView.StructureViewModel
import com.intellij.ide.structureView.TreeBasedStructureViewBuilder
import com.intellij.lang.PsiStructureViewFactory
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiFile
import me.salzinger.intellij.coco.psi.CocoFile

/**
 * Created by Thomas on 29/03/2015.
 */
class CocoStructureViewFactory : PsiStructureViewFactory {
    override fun getStructureViewBuilder(psiFile: PsiFile): StructureViewBuilder? {
        return if (psiFile !is CocoFile) {
            null
        } else object : TreeBasedStructureViewBuilder() {
            override fun createStructureViewModel(editor: Editor?): StructureViewModel {
                return CocoStructureViewModel(psiFile, editor)
            }

            override fun isRootNodeShown(): Boolean {
                return false
            }
        }

    }
}
